test/integration.test.js

```diff
@@ -1147,6 +1147,44 @@ describe('lint-staged', () => {
     // 'a/very/deep/file/path/file.js' matched '.lintstagedrc.json'
     expect(await readFile('a/very/deep/file/path/file.js')).toMatch('level-0')
   })
+
+  it('should support multiple configuration files with --relative', async () => {
+    // Add some empty files
+    await writeFile('file.js', '')
+    await writeFile('deeper/file.js', '')
+    await writeFile('deeper/even/file.js', '')
+    await writeFile('deeper/even/deeper/file.js', '')
+    await writeFile('a/very/deep/file/path/file.js', '')
+
+    const echoJSConfig = `module.exports = { '*.js': (files) => files.map((f) => \`echo \${f} > \${f}\`) }`
+
+    await writeFile('.lintstagedrc.js', echoJSConfig)
+    await writeFile('deeper/.lintstagedrc.js', echoJSConfig)
+    await writeFile('deeper/even/.lintstagedrc.cjs', echoJSConfig)
+
+    // Stage all files
+    await execGit(['add', '.'])
+
+    // Run lint-staged with `--shell` so that tasks do their thing
+    await gitCommit({ relative: true, shell: true })
+
+    // 'file.js' is relative to '.'
+    expect(await readFile('file.js')).toMatch('file.js')
+
+    // 'deeper/file.js' is relative to 'deeper/'
+    expect(await readFile('deeper/file.js')).toMatch('file.js')
```

Comment:
I wonder if that's the correct behavior. I can imagine a situation when there are multiple config files discoverable but only one root `node_modules` exist. How this change will affect it then? Or is it only due to `relative: true`?

Reply:
I think I get it now and I think it's fine. One more thing I'm wondering if this then should also affect the order in which tasks are run.

Reply:
Files are grouped exclusively to a single config, so the config groups should be safe to run in parallel. The same file can't be in multiple configs at the same time.

Reply:
Hmm, why not?

What if I want to run one check from the root of the monorepo and an additional check from sub-directory? I think it's a real use case and I'm not sure why we shouldn't support it.

Reply:
You mean run lint-staged checks twice? Hmm, not sure how to do that... maybe with `lint-staged --cwd .` and `lint-staged --cwd ./deeper`?

Reply:
Hmm I was thinking actually about a use case where I would have two configs defined, each one of them would have a same glob but different tasks. But I think it is how it will work now anyways?

Reply:
Yes, assuming the configs are in sibling directories and not parent/child. In the latter case only the deeper config would apply, because it's "closer" to a single staged file.

So each configuration file creates a boundary for all staged files inside the directory of the config. Placing a deeper config file will cut its own boundary and exclude the files from a parent config.

Reply:
Ouch I think that's not what I'd expect then. Why do you think this limitation is needed?

Reply:
This is a lot simpler, the other way would require running nested configs sequentially on a single file. I wonder if `lilconfig` supports finding multiple configs... currently it stops at the first.

Reply:
Hmm, although if I want to run two config files against a single staged file, I can do it like this:

```
npx lint-staged && npx lint-staged --config ./lint-stagedrc.json
```

Assuming I have configs at `./.lint-stagedrc.json` and `./packages/frontend/.lint-stagedrc.json`, and I stage `packages/frontend/some-file.js`.

The first run will auto-discover the closest config at `./packages/frontend/.lint-stagedrc.json`, and the second run will disable discovery and just use the explicit one. If both configs have `*.js` or similar, they will both match the file.

Reply:
IMHO the nested config runs multiple times is a very very error-prone design. Please dont do this...

Instead of detecting by lint-staged, nested config should be explicitly “merged” in user side.

For example,

tsconfig.json

`”extends”: “../../tsconfig.base.json”`

@okonet @iiroj thoughts?

Reply:
We have by design made the config structure simple, so it's only single-level like `Record<GlobPattern, string | string[] | ((files: string[]) => string | string[])`, so adding special fields is out of scope.

Reply:
@iiroj I think it's fine although I still think that one file could be included in multiple configs. This can be done as a follow up I think. Let's merge this one and I'll try to find some time next week to check it out. WYT?

Reply:
@chengcyber What do you mean by your comment? I'm not talking about merging configs but rather glob matching mechanics here.

Reply:
@okonet I think the only difficulty there is building "chains" of configs that should be run serially, vs unrelated configs that are safe to run in parallel. I would say deepest config runs first, right?

Since it's totally optional to have single/multiple configs, supporting both doesn't really cause any problems in the "isolated" monorepo case, since you can just _not_ put a config in the repo root.